src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -123,51 +169,48 @@ func (vmi *VMInterpreter_I) ApplyMessage(inTree st.StateTree, message msg.Unsign
 		),
 	)
 
-	if !sendRet.ExitCode().AllowsStateUpdate() {
-		// error -- revert all state changes -- ie drop updates. burn used gas.
-		outTree = inTree
-		outTree = _withTransferFundsAssert(
-			outTree,
-			message.From(),
-			addr.BurntFundsActorAddr,
-			_gasToFIL(sendRet.GasUsed(), message.GasPrice()),
-		)
-	} else {
-		// success -- refund unused gas
-		outTree = sendRetStateTree
-		refundGas := message.GasLimit() - sendRet.GasUsed()
-		TODO() // TODO: assert refundGas is nonnegative
-		outTree = _withTransferFundsAssert(
-			outTree,
-			addr.BurntFundsActorAddr,
-			message.From(),
-			_gasToFIL(refundGas, message.GasPrice()),
-		)
+	ok = _vmiUseGas(sendRet.GasUsed())
+	if !ok {
+		panic("Interpreter error: runtime execution used more gas than provided")
 	}
 
-	outTree, err = outTree.Impl().WithIncrementedCallSeqNum(message.To())
-	if err != nil {
-		// TODO: if actor deletion is possible at some point, may need to allow this case
-		panic("Internal interpreter error: failed to increment call sequence number")
+	ok = _vmiUseGas(gascost.OnChainReturnValue(sendRet.ReturnValue()))
+	if !ok {
+		return _applyError(compTreePreSend, exitcode.OutOfGas, addr.BurntFundsActorAddr, minerAddr)
+	}
+
+	compTreeRet := compTreePreSend
+	if sendRet.ExitCode().AllowsStateUpdate() {
+		compTreeRet = compTreePostSend
 	}
 
-	// reward miner gas fees
-	outTree = _withTransferFundsAssert(
-		outTree,
+	// Refund unused gas to sender.
+	refundGas := vmiGasRemaining
+	compTreeRet = _withTransferFundsAssert(
+		compTreeRet,
 		addr.BurntFundsActorAddr,
-		minerAddr, // TODO: may not exist
-		_gasToFIL(sendRet.GasUsed(), message.GasPrice()),
+		message.From(),
+		_gasToFIL(refundGas, message.GasPrice()),
 	)
 
-	return outTree, sendRet
+	return _applyReturn(
+		compTreeRet, vmr.InvocOutput_Make(sendRet.ReturnValue()), sendRet.ExitCode(),
+		addr.BurntFundsActorAddr, minerAddr)
```

Comment:
Blocking: the miner address is the wrong place to send funds. The fees go to the miner actor's owner address (along with the block reward).